optimize_workspace.sh

```diff
@@ -18,29 +18,61 @@ cargo --version
 rm -f target/wasm32-unknown-unknown/release/*.wasm
 
 # Build artifacts
-echo "Building artifacts in workspace ..."
+echo "Building artifacts in workspace..."
 /usr/local/bin/build_workspace
 
 mkdir -p artifacts
-echo "Creating intermediate hashes ..."
-sha256sum -- target/wasm32-unknown-unknown/release/*.wasm | tee artifacts/checksums_intermediate.txt
-
-echo "Optimizing artifacts in workspace ..."
+echo "Optimizing artifacts in workspace..."
 TMPARTIFACTS=$(mktemp -p "$(pwd)" -d artifacts.XXXXXX)
 # Optimize artifacts
 (
   cd "$TMPARTIFACTS"
 
   for WASM in ../target/wasm32-unknown-unknown/release/*.wasm; do
-    NAME=$(basename "$WASM" .wasm)${SUFFIX}.wasm
-    echo "Optimizing $NAME ..."
-    wasm-opt -Os "$WASM" -o "$NAME"
+    BASENAME=$(basename "$WASM" .wasm)
+    NAME=${BASENAME}${SUFFIX}
+    OPTIMIZED_WASM=${NAME}.wasm
+
+    INTERMEDIATE_SHA=$(sha256sum -- "$WASM" | sed 's,../target,target,g')
+    INTERMEDIATE_SHAS="../artifacts/checksums_intermediate.txt"
+
+    SKIP_OPTIMIZATION=false
+    if test -f "../artifacts/${OPTIMIZED_WASM}"; then
+      INTERMEDIATE_CACHE_HIT=$(
+        grep -Fxsq "$INTERMEDIATE_SHA" "$INTERMEDIATE_SHAS"
+        echo $?
+      )
+      OPTIMIZED_SHA=$(sha256sum -- "../artifacts/$OPTIMIZED_WASM" | sed 's,../artifacts/,,g')
+      OPTIMIZED_SHAS="../artifacts/checksums.txt"
+      OPTIMIZED_CACHE_HIT=$(
+        grep -Fxsq "$OPTIMIZED_SHA" "$OPTIMIZED_SHAS"
+        echo $?
+      )
+      if [ "$INTERMEDIATE_CACHE_HIT" -eq 0 ] && [ "$OPTIMIZED_CACHE_HIT" -eq 0 ]; then
+        SKIP_OPTIMIZATION=true
+      fi
+    fi
+
+    if [ "$SKIP_OPTIMIZATION" = true ]; then
+      echo "$BASENAME unchanged. Skipping optimization."
+    else
+      if test -f "$INTERMEDIATE_SHAS"; then
+        echo "Updating intermediate hash for ${BASENAME}..."
+        grep -v "$BASENAME" "$INTERMEDIATE_SHAS" >tmp_shas && mv -f tmp_shas "$INTERMEDIATE_SHAS"
+      else
+        echo "Creating intermediate hash for ${BASENAME}..."
+      fi
+
+      echo "$INTERMEDIATE_SHA" | tee -a "$INTERMEDIATE_SHAS" >/dev/null
```

Comment:
```suggestion
      echo "$INTERMEDIATE_SHA" >>"$INTERMEDIATE_SHAS"
```

Reply:
> Fixed, but with an addition. See: [4daeaef#diff-775370cff7b90a1fab3f86f20eceaa460788032dbcc72840d6dee16dab13af91R62-L66](https://github.com/CosmWasm/rust-optimizer/commit/4daeaef18be1afd14d48f66de231011a617a57a0#diff-775370cff7b90a1fab3f86f20eceaa460788032dbcc72840d6dee16dab13af91R62-L66)

That's not really needed, as `>>` works for "appending" to non existing files as well.

Reply:
Ah yeah true. Simplified, pushed